### ISSUE TITLE
After migrating to 0.9.12, Getting exception : org.reflections.RflectionException: Scanner SubTypeScanner was not configured , even after configuring the scanner.

### ISSUE THREAD

**Issue by anandmnair** [ronmamo/reflections#273]
I was using 0.9.11 and with scanners it was working fine.  but by just changing the version to 0.9.12 , stared getting the exception for "org.reflections.RflectionException: Scanner SubTypeScanner was not configured".

```java
Reflections reflections = new Reflections(new ConfigurationBuilder()
     .setUrls(ClasspathHelper.forPackage("my.project.prefix"))
     .setScanners(new SubTypesScanner(), new TypeAnnotationsScanner()),
     .filterInputsBy(new FilterBuilder().includePackage("my.project.prefix"))
     );

reflections.getSubTypesOf(someClass);  // here it throws exception 

```

I noticed that, when I add a class which extend some class to the package, this error goes. 
Same with FieldAnnotationScanner , or any other scanner. 

To try , please provide an invalid package for reflection creation, you will see all these errors coming even after configuring all scanners with reflection

**Comment by azell** [comment]
Having an `enum` type in the package path will also fix the problem, as it will force `SubTypesScanner` to be invoked.

**Comment by aheritier** [comment]
Same problem for me. Not sure if it's a regression of if I should adapt some code

**Comment by azell** [comment]
Think removing `store.getOrCreate` method is the culprit.

**Comment by koppor** [comment]
Same issue for `ResourcesScanner` (e.g., example from 
https://github.com/ronmamo/reflections/blob/gh-pages/UseCases.md#find-resources-in-your-classpath)

Or

```java
    Reflections reflections = new Reflections(new ResourcesScanner());
    var resources = reflections.getResources(Pattern.compile(".*pom\\.xml"));
```

    Exception in thread "main" org.reflections.ReflectionsException: Scanner ResourcesScanner was not configured

Does not happen with 0.9.11.

**Comment by chia7712** [comment]
> Think removing store.getOrCreate method is the culprit.

+1

```java
    public Reflections(final Configuration configuration) {
        this.configuration = configuration;
        store = new Store(configuration);

        if (configuration.getScanners() != null && !configuration.getScanners().isEmpty()) {
            //inject to scanners
            for (Scanner scanner : configuration.getScanners()) {
                scanner.setConfiguration(configuration);
                scanner.setStore(store.getOrCreate(scanner.getClass().getSimpleName())); // this line
            }

            scan();

            if (configuration.shouldExpandSuperTypes()) {
                expandSuperTypes();
            }
        }
    }
```

In 0.9.11, the constructor adds collection of all configured scanners to Store before does scan classes. By contrast, 0.9.12 create collection of all configured scanners only if it does scan something successfully. Personally, the error "was not configured" happening in 0.9.12 is a litter weird since the scanner is indeed "configured".

Could we keep the behavior of 0.9.11 (pre-create collection for all configured scanners) to fix this issue ?

**Comment by tmortagne** [comment]
Same issue here with ResourcesScanner.

Works fine when resource are found but crashing when there is nothing matching the configuration.
Looks like when no resource is found matching the configuration the corresponding map is null instead of being empty which make `Store#get(String)` think (wrongly) that this scanner was not even configured.

**Comment by GodLiang** [comment]
I also have this problem.I wan to change another lib.

**Comment by porunov** [comment]
We just switched to https://github.com/aschoerk/reflections8 (version 0.11.7). Even so it says that it is obsolete after reflections 0.9.12 release, they don't have this bug.
The fix to this issue is provided here: https://github.com/ronmamo/reflections/pull/278 but it wasn't reviewed / merged by repository maintainers.

**Comment by tr7zw** [comment]
What's the status here? This is a breaking issue that's been open for 5-6 months now, and the pr to fix it got closed without an update?

**Comment by TanVD** [comment]
Any updates? It is really strange that PR with fix of a problem was closed even without a comment.

**Comment by mfvanek** [comment]
I've faced with the same bug. Migrated to reflections8

**Comment by dano** [comment]
I haven't tested it to confirm, but it looks like @ronmamo might have fixed this via https://github.com/ronmamo/reflections/pull/275 on April 26th, the same day the PR that someone provided here was closed without comment.

**Comment by AntonOyung** [comment]
I upgraded 2 days ago form 0.9.11 to 0.9.12 and I am still seeing this issue

**Comment by dano** [comment]
Just to clarify my previous comment - you'd have to be using 0.9.13-SNAPSHOT to have the commit I think may have fixed it.

**Comment by AntonOyung** [comment]
How can I get 0.9.13-SNAPSHOT?

**Comment by dano** [comment]
@AntonOyung I think you'll have to build it from source yourself. It doesn't look like SNAPSHOT builds get pushed to a maven repository.

**Comment by xresch** [comment]
I ran into the same issue when migrating from JDK8 to JDK11.
For me the issue didn't occur anymore when I removed the configuration builder and just instantiated the Reflections class like below.
This way works for me for JDK 8, 11 and 13  and with reflection version 0.9.12.

```java
//-------------------------------------------
// Old instantiation that throws an error with 
// reflections 0.9.12 and JDK 11 or higher

//       Reflections reflections = new Reflections(new ConfigurationBuilder()
//            //.filterInputsBy(new FilterBuilder().exclude(FilterBuilder.prefix("com.xresch.cfw.")))
//            .filterInputsBy(new FilterBuilder().exclude(FilterBuilder.prefix("java.")))
//            .setUrls(ClasspathHelper.forClassLoader())
//            //.setScanners(new SubTypesScanner(), new TypeAnnotationsScanner())
//       );
	
//-------------------------------------------
// New instantiation that works with 
// Reflections 0.9.12 and JDK8 or higher
// scans all classes	
Reflections reflections = new Reflections("");

//-------------------------------------------
// No changes here
Set<Class<?>> types = reflections.getTypesAnnotatedWith(CFWExtensionApplication.class);

for(Class<?> clazz : types) {
   if(CFWAppInterface.class.isAssignableFrom(clazz)) {
	   new CFWLog(logger).method("loadExtentionApplication").info("Load CFW Extension Application:"+clazz.getName());
	   
	  try {
		CFWAppInterface instance = (CFWAppInterface)clazz.newInstance();
		return instance;
	} catch (InstantiationException | IllegalAccessException e) {
		new CFWLog(logger).method("loadExtentionApplication").severe("Error loading CFW Extension Application:"+clazz.getName(), e);
	}
   }
}
```

**Comment by crykn** [comment]
This has been broken for over nine months by now... @ronmamo can we get a bugfix release 0.9.13 or is this library considered abandoned?

**Comment by Taironn** [comment]
It is still broken. Should I be building on this library, or look for alternatives?

**Comment by crykn** [comment]
Honestly, I think it is mind-boggling that someone "maintaining" a library with 3.4k stars is not even commenting on such a common issue. For the time being, we have switched to [reflections8](https://github.com/aschoerk/reflections8) as was suggested further up.

**Comment by tr7zw** [comment]
Seems like there is no maintainer active anymore, might be time to fork this repo and start merging the pull requests that started to pile up with release on maven central.

**Comment by l1ving** [comment]
This is a shame. A library that's this widely used should at least be handed over to someone else who wants to maintain it

**Comment by xresch** [comment]
Guys please be aware that we are on an open source platform and this developer did some free work so you don't have to do it yourself.
I also wait for a fix, but we do not know the circumstances why the developer is currently not as responsive as in the past(maybe he currently does a world trip or has personal issues, how knows).

There are currently 569 forks of this project, maybe someone has already found a fix or you can fork it yourself and investigate it. Or you can try https://github.com/aschoerk/reflections8 as suggested above.

**Comment by astubbs** [comment]
If you look at Romano's activity levels, it' seems quite odd https://github.com/ronmamo
Hope they're ok...

**Comment by astubbs** [comment]
Does anyone know how to contact them? Can't see anything on their profile page.

**Comment by ronmamo** [comment]
apologize for this inconvenience !

scanner was not configured exception - this is a known issue in 0.9.12, a simple workaround is to check if the `getStore()` contains index for the scanner before querying. [next version 0.10](https://github.com/ronmamo/reflections/pull/335) fixes this.

**Comment by ronmamo** [comment]
fixed on 0.10
